Add: Navigation menus to the browse mode sidebar.

jorgefilipecosta opened the pull request:
This PR adds navigation menus to the browse mode sidebar. Allowing the user to manage menus from there.
It moves the current navigation sidebar to the left.

cc: @jasmussen, @jameskoster 
## Screenhots
![image](https://user-images.githubusercontent.com/11271197/206742568-8d06e4f7-cec0-4055-bec9-f1fff5d26a9b.png)
![image](https://user-images.githubusercontent.com/11271197/206742619-af3d641f-57c0-4be1-b469-f8b08005624c.png)

## Follow up
There is a new navigation block manager being implemented at https://github.com/orgs/WordPress/projects/60/views/1. We should try to use that new component here.

mtias commented on packages/edit-site/src/components/sidebar-navigation-screen-main/index.js:
```diff
@@ -49,6 +49,14 @@ export default function SidebarNavigationScreenMain() {
+						withChevron
+						icon={ navigation }
+					>
+						{ __( 'Navigation menus' ) }
```
Let's use a singular here

jasmussen commented:
Outstanding, thank you! This GIF shows the new navigation item in the sidebar left of the new browse mode interface:

![nav](https://user-images.githubusercontent.com/1204802/207010015-6fd0bd68-56f2-4843-a01c-3947d41c6d8a.gif)

A few things. Let's use singular for menus. But beyond renaming it to "Navigation menu", can we just call it "Navigation"?

The other aspect is that ideally a click on a navigation item navigates to that page. So in the GIF above I have Blog, About, Gallery and Contact. While I should be able to rearrange items in the _menu_, I should also be able to click an item to go there in the browse view. We can think of it as the first iteration of actual browse mode.

Probably doesn't have to happen in this PR, but worth noting.

We can remove this item now:

<img width="341" alt="Screenshot 2022-12-12 at 10 32 08" src="https://user-images.githubusercontent.com/1204802/207010936-e654eb1d-aefb-4ea7-a03e-31a481adebfc.png">

Something is off about the positioning here:

<img width="340" alt="Screenshot 2022-12-12 at 10 33 34" src="https://user-images.githubusercontent.com/1204802/207011198-56fef437-0f4d-41b0-9ec7-30d0ed7ef6ec.png">

There's a missing focus style, so the default chrome focus style is applied:

<img width="330" alt="Screenshot 2022-12-12 at 10 34 14" src="https://user-images.githubusercontent.com/1204802/207011351-d1589465-7ff6-4de7-8e67-0ef3f099727b.png">

The alignment of icons here is a bit wrong:
<img width="109" alt="Screenshot 2022-12-12 at 10 35 00" src="https://user-images.githubusercontent.com/1204802/207011454-34bfa036-f9d5-416d-88ae-09f5af8ee6c7.png">

I wonder, to get the alignment right, can we use the same componentry as is used for templates?

<img width="288" alt="Screenshot 2022-12-12 at 10 35 19" src="https://user-images.githubusercontent.com/1204802/207011549-e05b7235-6781-4473-a225-4ef1abe2116e.png">

Nice one. Let me know if you need some help here with the intricacies.

jorgefilipecosta committed:
Add: Navigation to the browse mode sidebar.

jorgefilipecosta commented:
Hi, @jasmussen, and @mtias, the title was changed to navigation as suggested.
This PR was updated to use the off-canvas editor.
The look is:
![image](https://user-images.githubusercontent.com/11271197/210267075-53cb198a-3a1b-47cc-9a12-10bdd4c8a586.png)


Because the base component is not exposed, it requires the Off-canvas navigation editor feature to be enable at wp-admin/admin.php?page=gutenberg-experiments.


> The other aspect is that ideally a click on a navigation item navigates to that page. So in the GIF above I have Blog, About, Gallery and Contact. While I should be able to rearrange items in the menu, I should also be able to click an item to go there in the browse view. We can think of it as the first iteration of actual browse mode.

> Probably doesn't have to happen in this PR, but worth noting.

Clicking to go to the page will be a follow-up after this PR is merged.

Is there any update we should do before merging this PR?

jameskoster commented:
A couple of observations:

It would be good if the navigation panel items could be more dimensionally aligned with the Templates + Template Part panels. Hopefully this illustration explains what I mean:

<img width="296" alt="Screenshot 2023-01-04 at 17 42 40" src="https://user-images.githubusercontent.com/846565/210616838-443a485b-4d71-4162-bae1-49330be01063.png">

The menu dropdown is a bit noisy, would it be possible to style it to match the 'dark' theme?

It's a bit strange to have hover / focus styles on the items in the list when there is no on-click interaction. It feels a bit broken like this. I appreciate there's drag/drop but it's not obvious, perhaps a different cursor (`grab`?) on hover is worth a try.

In the ellipsis menu, the inspector toggle ("Show more settings") does nothing. Should we hide it? Locking feels a bit strange in this context, as do the create template part / reusable block actions. I appreciate these items might be something that needs to be fixed in the off-canvas nav component rather than here. I don't know if they should be blockers for this PR.

Most likely one for a follow-up, but should there be a way to create new menus here, similar to creating new templates? If so, I wonder if the add menu item should have a label? IE:

<img width="373" alt="Screenshot 2023-01-04 at 17 46 17" src="https://user-images.githubusercontent.com/846565/210617441-4bb24bb9-ef02-4603-b9f2-fe0d6e74b637.png">

youknowriad commented on packages/edit-site/src/components/sidebar-navigation-item/style.scss:
```diff
@@ -1,4 +1,5 @@
-.edit-site-sidebar-navigation-item.components-item {
+.edit-site-sidebar-navigation-item.components-item,
+.edit-site-sidebar-navigation-screen__content .edit-site-navigation-inspector .components-button {
```
Why do we need a specific style here in a generic component?

jorgefilipecosta replied:
I removed the styles for .edit-site-sidebar-navigation-item.components-item, the other styles are required so the buttons looked "themed" on the dark mode.

jorgefilipecosta commented:
Hi @jameskoster,

I fixed the margin issue. I updated the cursor to be grabbed.
I created a side PR to make the inserter of the off-canvas menu editor include the text "Add menu item," which when both PRs are merged, will provide the following UI:
![image](https://user-images.githubusercontent.com/11271197/211053055-c9de284a-0f1c-4e95-bf54-58040c6f3da5.png)


> Most likely one for a follow-up, but should there be a way to create new menus here, similar to creating new templates?

I'm not sure if we should offer the ability to create menus here. I even imagine that in the future, the menus available to browse may be restricted to the current menus used on the page/template (in most cases, just one menu is used, so the user only sees that menu). But my vision for the navigation section may be wrong.

I think the options available on the ellipsis menu can be solved in a separate PR. Like the "Add menu item" is. These are changes on the off-canvas editor itself.

jameskoster commented:
Thanks @jorgefilipecosta. I'll dig into this more tomorrow, but one thing I noticed is that the link styles in this PR are now different to trunk:

### Trunk


https://user-images.githubusercontent.com/846565/211064111-43876bd3-7e73-4446-a63e-703efe4209cf.mp4

### This PR


https://user-images.githubusercontent.com/846565/211064141-4df62e04-5638-457f-86f6-f60cb6b60ea7.mp4

---

This view isn't very helpful:

<img width="355" alt="Screenshot 2023-01-06 at 17 23 10" src="https://user-images.githubusercontent.com/846565/211064572-f177bf87-5805-47c2-b589-5141077e5704.png">

If we don't intend to allow the creation of menus here, perhaps we should hide the panel altogether when no menus are found?

---

Edit: I noticed the alignment and widths aren't right yet either. Ideally the icon should align roughly with the edge of the panel title, and reach row should be full width. Something like:

<img width="356" alt="Screenshot 2023-01-06 at 17 32 17" src="https://user-images.githubusercontent.com/846565/211066124-00caa149-6ddc-486c-a5d0-ae900009cc6b.png">

youknowriad commented:
This PR is looking good in general.

The main feedback for me when testing the PR is that now it's becoming more "urgent" to figure out the saving flows outside the frame: when you add a page to the menu, it's still unsaved by default. I'm happy to address separately if needed but we'd need some designs.

The other thing is that when clicking pages/posts on the navigation tree, I expected the corresponding template with the right post/page in context to be loaded in the frame and ideally the element should be marked "selected" in the tree (like we do for the template list), thought that last point is arguable as we could potentially "only" select the corresponding template as a first step.

jameskoster commented:
Saving in browse view needs some holistic consideration imo. Specifically; do we want a global save button that opens the multi-entity save panel, or do we want contextual save buttons in the panels that require them (Navigation and Styles)?

Alternatively, since creating a template automatically publishes it, we might make a case that any changes made in the browse view are auto-saved. 

I think a global save button (inside the site hub?) that opens the multi-entity save panel in a modal might work best, but I'd love to hear thoughts from others. Should we open a separate discussion for this?

> we could potentially "only" select the corresponding template as a first step.

I think it would be a little mysterious to see the frame update on click, but not display the content of the item you clicked on. So I'd lean towards tackling template+content together as a separate effort.

youknowriad commented:
> I think a global save button (inside the site hub?) that opens the multi-entity save panel in a modal might work best, but I'd love to hear thoughts from others. Should we open a separate discussion for this?

I agree it needs to be discussed and solved separately, so yeah let's go with a separate discussion here. I also lean towards a global save button that morphs into the editor's save button when moving to edit mode.

jasmussen commented:
Coming back to this one, just wanted to comment on this:

> The menu dropdown is a bit noisy, would it be possible to style it to match the 'dark' theme?

I agree, both in the navigation block and here, and I wonder if instead of a dropdown (`select`) it could be a title plus a chevron with a popover.

SaxonF commented:
Now that we have the ability to manage menus via the navigation block, do we really need a dedicated space for menu management? I'm not personally sure we do as we are leaning more in to direct manipulation of menus (e.g. click navigation block -> edit ) as opposed to managing menus out of context. What value do we get and is it worth it?  You can also edit menus in isolation via template parts / patterns. 

If the answer is definitely yes, at the very least the list view experience designed and built for the inspector should be re-used here.

jasmussen commented:
> Now that we have the ability to manage menus via the navigation block, do we really need a dedicated space for menu management?

One key value to hold is to consider the 80% use case. For example, 1 header, 1 menu, 1 footer. In those cases, being able to edit your menu right from the design section, I suspect, will bring with it a lot of ease of use.

We still need for multiple menu management, and menu management inside the editor, to be vastly simplified. The improvements happening in that space are exciting and needed. But those still require you to click "Edit" in context of a random page, meaning the global context of a menu is slightly lost. So in that light, I see having access to both styles and menus right in the design section as a win.

jorgefilipecosta commented:
Hi @jameskoster, thank you for this review.

> Thanks @jorgefilipecosta. I'll dig into this more tomorrow, but one thing I noticed is that the link styles in this PR are now different to trunk:

 Fixed 👍 

> Edit: I noticed the alignment and widths aren't right yet either. Ideally, the icon should align roughly with the edge of the panel title, and reach row should be full width. Something like:


I applied a fix to try to align things.
![image](https://user-images.githubusercontent.com/11271197/212710034-1eca5c4a-eb11-4127-b907-ac5119c131d0.png)

In most menus, we will have submenus, so we have an arrow to expand the submenu. That row is now aligned with the panel title. But the icon of the menu item is not, as shown in the image, because we need space for the arrow.

> If we don't intend to allow the creation of menus here, perhaps we should hide the panel altogether when no menus are found?


I applied the suggestion, and navigation is hidden if there are no menus.

jorgefilipecosta commented:
> Coming back to this one, just wanted to comment on this:

> > The menu dropdown is a bit noisy, would it be possible to style it to match the 'dark' theme?

> I agree, both in the navigation block and here, and I wonder if instead of a dropdown (select) it could be a title plus a chevron with a popover.

Hi @jasmussen, I guess we could use the CustomSelectControl component, so we have a "select," and then we could style it to match what we want? From the a11y point of view, we would take advantage of all the work done and the control would still look like a select.

youknowriad commented:
Test this one again, for me the important missing piece right now is the ability to click a page/post to view/edit it.

jasmussen commented:
> I guess we could use the CustomSelectControl component, so we have a "select," and then we could style it to match what we want? From the a11y point of view, we would take advantage of all the work done and the control would still look like a select.

Actually can we try just locking it, as a starting point, to a single menu? That is to say, the first menu saved. If you need to edit other menus, you have to go into the editor, find that menu, and edit there? This is meant to be the _global_ menu, so it might be good to lock it to that.

jorgefilipecosta commented:
> > I guess we could use the CustomSelectControl component, so we have a "select," and then we could style it to match what we want? From the a11y point of view, we would take advantage of all the work done and the control would still look like a select.
> 
> Actually can we try just locking it, as a starting point, to a single menu? That is to say, the first menu saved. If you need to edit other menus, you have to go into the editor, find that menu, and edit there? This is meant to be the _global_ menu, so it might be good to lock it to that.

Done. I removed the select from the navigation sidebar. We have a smart logic to detect what menu appears, which works well.

jameskoster commented:
Thanks @jorgefilipecosta. The styling is better but a few details remain:

Specifically the rows are too narrow, and the ellipsis existing outside of the hover area looks a bit strange.

<img width="357" alt="Screenshot 2023-01-17 at 14 22 49" src="https://user-images.githubusercontent.com/846565/212923592-7125ac26-5ecf-408d-8649-16d00914ff6b.png">

Minor point: the submenu icon looks a bit odd. I don't think we need that – can we just use the regular icon?

It would be really great to align menu item styling across all panels:

<img width="334" alt="Screenshot 2023-01-17 at 14 29 55" src="https://user-images.githubusercontent.com/846565/212925281-d5cbfd7b-959a-4dbb-a017-cde8ec6364d8.png">

Here's the [Figma](https://www.figma.com/file/wXU0x4xZUuiNmLrtCWfFZW/Browse-view?node-id=2067%3A109207&t=NYLYVUqujfH3vlzG-1) file in case it's helpful.

jorgefilipecosta commented:
Hi @jameskoster,
I did another round of updates, trying to go in the direction you are proposing. That's how it looks currently:
![image](https://user-images.githubusercontent.com/11271197/213169169-26c62a2d-2f7a-4071-9079-f57ed4751f97.png)

Let me know if any other updates are necessary.

> Minor point: the submenu icon looks a bit odd. I don't think we need that – can we just use the regular icon?

Regarding this point, we are using the icon of the submenu block, because the menu is just the new offsite navigation editor. Changing the icon here would change it everywhere including the list view, block inspector, and the block toolbar. Maybe we can think of a new icon that fits all these cases?

jameskoster commented:
Thanks for your perseverance, that looks _much_ better :) 

> Changing the icon here would change it everywhere including the list view, block inspector, and the block toolbar.

Fair enough, let's tackle that separately.

youknowriad commented on packages/edit-site/src/components/sidebar-navigation-screen-main/index.js:
```diff
@@ -15,12 +17,33 @@ import SidebarNavigationScreen from '../sidebar-navigation-screen';
 import SidebarNavigationItem from '../sidebar-navigation-item';
 
 export default function SidebarNavigationScreenMain() {
+	const { navigationMenus } = useSelect( ( select ) => {
```
I've noticed that when I open the site editor, I see two menu items (templates and template parts) and a few milliseconds later, I see the navigation one. I wonder if we should use `useSuspenseSelect` here and add a `Suspense` provider around the sidebar to show a loader or something before the navigation menus are loaded. (We've never used suspense before so this might not be ready yet but I think worth a try, could be a follow-up)

youknowriad reviewed (approved):
LGTM 👍 

I think it would be good to follow-up with the "browse mode" (click posts and pages to open them in the site editor) cause I feel it's the main added value for moving this navigation editor to the sidebar.

jasmussen commented:
> I think it would be good to follow-up with the "browse mode" (click posts and pages to open them in the site editor) cause I feel it's the main added value for moving this navigation editor to the sidebar.

Yep, this is a pretty crucial piece!

youknowriad commented:
Oh something I noticed here, should we remove the right navigation sidebar when you're in "edit mode". I think it's redundant now and I think this was always a temporary place for it.

jasmussen commented:
> Oh something I noticed here, should we remove the right navigation sidebar when you're in "edit mode". I think it's redundant now and I think this was always a temporary place for it.

Yes!

ndiego commented:
@jorgefilipecosta @jasmussen I was testing this PR with templates that have both header and footer navigation. The Navigation panel displays the menu in the header, which looks to be the expected result. 

I think this might lead to some confusion, at least it did for me 😅. Since Template and Template Parts allow you to edit all templates and parts, I assumed that I would be able to see and edit all navigation menus within the browse mode Navigation panel. I know that menus can also be managed within the Editor, but this abstracted view is fantastic and feels a lot like the classic menu management screens of old, which I know many people love—food for thought.

I also ran into the issue noted [here](https://github.com/WordPress/gutenberg/pull/47354#issuecomment-1401651146), that saving a modified menu still requires the user to go into Edit mode and then save.

mtias commented:
Yeah, we'll likely add the dropdown back at some point, but in a way that is less obtrusive or disorienting than the large white dropdown at the top. It's encouraging to hear this feels like an intuitive interface to manage things at a high level :)
